### DISCUSSION TITLE
Feature Request: Add MQTT5 support including Request-Response Pattern support

### DISCUSSION THREAD

**Discussion by Snapstromegon** [arendst/Tasmota#17969] [Ideas]
**Problem:**

Building an external API interface (like a TasmotaAdapter library) currently is pretty hard when you want to use MQTT, because you can't send a command and get a response.

**Possible Solution:**

MQTT5 introduces support for Request-Response patterns using `Response Topics` and `Correlation Data` which allows a command issuer to know what message is a direct response to the command and avoids race conditions.

**Suboptimal Alternatives:**

_Changing the API_

Aside from supporting the "standard" MQTT5 Request/Response pattern, an alternative that implements this on top of MQTT3 is also possible, but would significantly change the way the current way of issueing commands and receiving results works and would probably either require to break the existing API (which is really bad) or to add a new MQTT API that specifically allows issueing commands with correlation data (which bloats the API surface and is also suboptimal).

_Using the HTTP API_

Using the HTTP API is not always possible, because most web services (e.g. when using a webapp to issue commands) are behind a HTTPS endpoint and therefore browsers won't allow access to HTTP APIs from those apps.
Also in my usecase I want to use the MQTT log to see which commands got issued when to which device. This kind of monitoring is not possible with the HTTP API (outside of fairly dirty workarounds using the MqttLog feature).

_Using Berry Scripting_

Scripting is disabled in default builds and is not even possible on all devices because of size constraints. This means that if someone wants to use such an adapter library, it's fairly likely that they have to reflash tasmota for it to work and to "install" the right scripts on it.

---

I'm open to other suggestions and normally I would offer to chime in with a contribution, but I don't feel like I would be able to implement a working solution here.

For reference: This started as a message in discord here: https://discord.com/channels/479389167382691863/790182550085304342/1076105810788233296

**Comment by sfromis** [comment]
As I already suggested on Discord, such features are primarily up to the MQTT library used by Tasmota. This is currently [pubsubclient](https://github.com/knolleary/pubsubclient) which can be MQTT 3.1.1 or 3.1. Not MQTT 5.

Technical detail: Berry is by default enabled on ESP32 builds, but "scripting" is an entirely different feature, not included in default builds, but also supporting ESP8266.

**Reply by Snapstromegon** [reply]
Thanks for the reply.
I searched for "MQTT" in the codebase and because of that didn't find pubsubclient.
They already have an open issue since 2018 for MQTT5 support: https://github.com/knolleary/pubsubclient/issues/531

The question now is, wether Tasmota wants to wait for MQTT5 support in PubSubClient, build an own "workaround"/solution, or just postpone this idea for now.

In my opinion waiting for MQTT5 support in PubSubClient is the best aproach even though it means that this gets potentially blocked for a long time.

**Reply by barbudor** [reply]
I see moving from asynchronous to synchronous a walkback.
At first sight it makes programming easier but it's worse in term of threading, resource management.
In my business we are using Kafka, SQS, Lambda. For me, all modern computing infrastructure are based on asynchronous.

**Reply by sfromis** [reply]
The tricky thing about adding some token (to be included in responses) to the existing commands is mostly that many messages published by Tasmota are not direct responses to a particular command, even in cases where they are indirectly caused by a command being processed. This is due to Tasmota also internally being asynchronous about processing things happening, instead of simple request/response flows.

Thus, even if MQTT 5 support in pubsubclient arrived tomorrow (or getting MQTT 5 support some other way), it would not be "just so" to restructure Tasmota to have a strong correlation from commands to what is published for various reasons, not limited to commands.

**Reply by Snapstromegon** [reply]
I do not think of request/response as "synchronous", but I see where this is coming from.
In the case where I actively want to do a RPC (which executing a command in my opinion is), it's still valuable to which command a result refers to.

Right now (as I see it), it's not easily possible to actively make decision based on the state of a tasmota device without having a whole replicated state of it in store (which can move out of sync).
In general I'm also a huge fan of things like the CQRS pattern, but some things just need to associate a response to a request.

To me right now it's a little like making a DB request from a lambda function to search for a user and at some point you get a response with a user, but you don't know if the response is for your search or not.

An easy way to have a duplicated state store move out of sync is when messages get reordered on the way to or from the broker.

**Reply by sfromis** [reply]
Well, the relevant analogy instead of your db request case would be a device with 5 relays. When a relay changes state, you will get to know which relay changed state to what, but not if it was "you" causing it to change state. This is very different from not knowing if a "user" response is relevant. Thus, in the real world you generally have what you need to make a decision based on the *state* of the device.

This was not meant as hairsplitting about words, more that things are more straightforward with simple cases like controlling relays, instead of database requests. In other words, I think that you may be on the path of premature optimization.

**Comment by Jason2866** [comment]
There will be no support for mqtt 5.x in near future.

**Comment by s-hadinger** [comment]
Although for esp32, esp-idf supports MQTT5, it would be a major refactoring. It does not seem worth it for now.

**Comment by winksaville** [comment]
It's now going on 3 years since this particular discussion, what are the recent thoughts on MQTT5 support?

**Reply by arendst** [reply]
No change.

**Comment by s-hadinger** [comment]
Mqtt 5 would mean deep changes in Tasmota and most certainly regression. I haven’t seen any killer feature in mqtt 5 that would make the change worth the risk

**Comment by Steve-Mcl** [comment]
Have alternative MQTT v5 capable libs been considered guys?

There are a few now. e.g: 
* [mqtt5nano](https://docs.arduino.cc/libraries/mqtt5nano/)
* [ESP32_MQTTv5](https://www.arduinolibraries.info/libraries/esp32_mqt-tv5)
* [H4AsyncMQTT](https://github.com/HamzaHajeir/H4AsyncMQTT)
* And more I suspect!

I only resurrect this thread as I am looking to add core "MQTT Request" support to Node-RED (ref: https://discourse.nodered.org/t/we-have-http-request-tcp-request-serial-request-any-objections-to-an-mqtt-request/101413) and one of our users asked how/if Tasmota could be supported.  This would go very nicely hand in hand with Tasmota me thinks!

Keep up the good work 👊

**Reply by sfromis** [reply]
Are you asking for features to be added to Tasmota to actually take advantage of the new options with MQTT 5? Until that, it should make no difference to use a MQTT 5 capable lib, but would likely cost flash space. As long as the broker supports both levels, you can mix clients.

To fully support request-response is not really about getting a MQTT 5 capable lib (that's indeed the easy part), but about restructuring Tasmota to even have the concept of strongly linking a publish to whether it was the result of receiving a specific payload on a subscription, vs other actions taking place at the same time.

In simpler cases, the response is synchronous, and could be linked to the command name, but still not the specific request if multiple of same command name are running at the same time. With some commands the results are asynchronous, could be multiple lines without an explicit end marker, and generated after the processing of the command was completed. Same results as from a command could also be generated without a command being executed, like triggered by local automations.

**Reply by Steve-Mcl** [reply]
> Are you asking for features to be added to Tasmota to actually take advantage of the new options with MQTT 5?

No, only echoing sentiment and tipping a wink to progress in Node-RED that would compliment any support for v5 request-response support in Tasmota.

As for internal implementation (be it sync or async or however you handle long running requests with multiple incoming requests) is not something I can comment on I'm afraid (I dont know the Tasmota code base) - I am just the messenger here 👼

**Reply by s-hadinger** [reply]
The logic of running a command synchronously and returning the response already exists for Web requests, but indeed plugging it into MQTT would require some core changes in Tasmota. To be honest, I'm more concerned about possible side effects of changing the core MQTT lib.

[H4AsyncMQTT](https://github.com/HamzaHajeir/H4AsyncMQTT) seems to rely on ESPAsyncWebserver that we don't use because of stability issues.
[ESP32_MQTTv5](https://www.arduinolibraries.info/libraries/esp32_mqt-tv5) looks nice but it is coded as a single C file with ZERO comments (seriously???)
[mqtt5nano](https://docs.arduino.cc/libraries/mqtt5nano/) on the other side is heavy weight and includes its own crypto which would have a huge impact on code size.

From the list above, [ESP32_MQTTv5](https://www.arduinolibraries.info/libraries/esp32_mqt-tv5) would be the best candidate but the lack of any comments makes me suspicious regarding code quality and maintainability

**Reply by sfromis** [reply]
While the coupling between command and response does exist for simple commands, this is leaky, like not working with `Backlog`, asynchronously generated messages like `WifiScan` responses, or when command execution involves a timer. Hence, request-response would not be quite reliable.

**Reply by Steve-Mcl** [reply]
I think those long running tasks could simple be responded immediately with an acknowledgement that the "scan has been scheduled" & perhaps indicate what topic they will be published on? Obvious continuous things like logs may not make sense - so don't offer a request-response option for those things?

Please note, this is 100% off the top of my head so forgive me if it's utter nonsense or doesn't align with the product vision or standards.  My only reason for dredging this up was to inform on the node-red progress for if/when ever you decide (or not) to support mqtt V5 (specifically request-response).

**Reply by s-hadinger** [reply]
Well I do agree that request/response in MQTT is not trivial, but I don't know the details of MQTT5 and how simpler it gets. Anyways this would be a heavyweight change and it is not really in our plans right now.
